forge-ai/src/main/java/forge/ai/AiAttackController.java

```diff
@@ -511,6 +508,82 @@ public boolean apply(final Card c) {
         return notNeededAsBlockers;
     }
 
+    public void reinforceWithBanding(final Player player, final Combat combat) {
+        reinforceWithBanding(player, combat, null);
+    }
+
+    public void reinforceWithBanding(final Player player, final Combat combat, final Card test) {
+        List<String> bandsWithString = Arrays.asList("Bands with Other Legendary Creatures",
+                "Bands with Other Creatures named Wolves of the Hunt",
+                "Bands with Other Dinosaurs");
+
+        List<Card> bandingCreatures = new CardCollection();
+
+        if (test == null) {
+            bandingCreatures.addAll(notNeededAsBlockers(player.getCreaturesInPlay()));
```

Comment:
Should reorder the calls here a bit by their computation costs:

- KW first
- switch isAttacking and canAttack
- `notNeededAsBlockers` probably last as most expensive

Reply:
Done :)
I also wanted to do notNeededAsBlockers last, but sadly, that method returns all the creatures that aren't needed as blockers, effectively negating any filtering...

Reply:
The way it's currently called is probably problematic:
it expects to be passed the potential attackers BUT at this point any number of them are already declared and unavailable for blocking anyway
so AI could greatly overestimate its defensive power when it grants the banders

Reply:
or maybe it will get filtered somewhere deeper
I'll investigate that as followup I think

Reply:
Ok it's been a while since I wrote that function and it's rather complex but I think it should behave correctly now
Would need more time for sanity checks later though 😴

Reply:
Nice, thanks a lot for help here! I played a few games and things seemed normal, but feel free to test additionally as well when able :)